fdbrpc/AsyncFileChaos.actor.h

```diff
@@ -26,94 +26,103 @@
 	void delref() override { ReferenceCounted<AsyncFileChaos>::delref(); }
 
-	uint8_t toggleNthBit(uint8_t b, uint8_t n) {
-		auto singleBitMask = uint8_t(1) << (n);
-		return b ^ singleBitMask;
-	}
+	static double getDelay() {
+		double delayFor = 0.0;
+		auto res = g_network->global(INetwork::enDiskFailureInjector);
+		if (res) {
+			DiskFailureInjector* delayInjector = static_cast<DiskFailureInjector*>(res);
+			delayFor = delayInjector->getDiskDelay();
 
-	void flipBits(void* data, int length, double percentBitFlips) {
-		auto toFlip = int(float(length*8) * percentBitFlips / 100);
-		TraceEvent("AsyncFileFlipBits").detail("ToFlip", toFlip);
-		for (auto i = 0; i < toFlip; i++) {
-			auto byteOffset = deterministicRandom()->randomInt64(0, length);
-			auto bitOffset = uint8_t(deterministicRandom()->randomInt(0, 8));
-			((uint8_t *)data)[byteOffset] = toggleNthBit(((uint8_t *)data)[byteOffset], bitOffset);
+			// increment the metric for disk delays
+			if (delayFor > 0.0) {
+				auto res = g_network->global(INetwork::enChaosMetrics);
+				if (res) {
+					ChaosMetrics* chaosMetrics = static_cast<ChaosMetrics*>(res);
+					chaosMetrics->diskDelays++;
+				}
+			}
 		}
+		return delayFor;
 	}
 
 	Future<int> read(void* data, int length, int64_t offset) override {
-		double delay = 0.0;
-		auto res = g_network->global(INetwork::enDiskFailureInjector);
-		if (res)
-			delay = static_cast<DiskFailureInjector*>(res)->getDiskDelay();
-		TraceEvent("AsyncFileChaosRead").detail("ThrottleDelay", delay);
-		return delayed(file->read(data, length, offset), delay);
+		double diskDelay = getDelay();
+
+		// Wait for diskDelay before submitting the I/O
+		// Template types are being provided explicitly because they can't be automatically deduced for some reason.
+		return mapAsync<Void, std::function<Future<int>(Void)>, int>(
+		    delay(diskDelay), [=](Void _) -> Future<int> { return file->read(data, length, offset); });
 	}
 
 	Future<Void> write(void const* data, int length, int64_t offset) override {
-		double delay = 0.0;
 		char* pdata = nullptr;
+
+		// Check if a bit flip event was injected, if so, copy the buffer contents
+		// with a random bit flipped in a new buffer and use that for the write
 		auto res = g_network->global(INetwork::enBitFlipper);
 		if (res) {
-			auto percentBitFlips = static_cast<BitFlipper*>(res)->getPercentBitFlips();
-			if (percentBitFlips > 0.0) {
-				TraceEvent("AsyncFileCorruptWrite").detail("PercentBitFlips", percentBitFlips);
-				pdata = new char[length];
+			auto bitFlipPercentage = static_cast<BitFlipper*>(res)->getBitFlipPercentage();
+			if (bitFlipPercentage > 0.0) {
+				pdata = (char*)arena.allocate4kAlignedBuffer(length);
 				memcpy(pdata, data, length);
-				flipBits(pdata, length, percentBitFlips);
-				auto diff = memcmp(pdata, data, length);
-				if (diff)
-					TraceEvent("AsyncFileCorruptWriteDiff").detail("Diff", diff);
+				if (deterministicRandom()->random01() < bitFlipPercentage) {
```

Comment:
There is no need to copy the buffer if no change is being made.  Copying the buffer is significant overhead.

Reply:
Done